src/base/notifications/request.cpp

```diff
@@ -0,0 +1,164 @@
+#include <QDebug>
+
+
+Notifications::Request::Request()
+    : m_title(QObject::tr("Notification"))
+    , m_message(QObject::tr("Notification from qBittorrent"))
+    , m_category(Category::Generic)
+    , m_urgency(Urgency::Normal)
+    , m_severity(Severity::No)
+    , m_torrentHash(BitTorrent::InfoHash())
+    , m_timeout(-1)
+{}
+
+QString Notifications::Request::title() const
+{
+    return m_title;
+}
+
+QString Notifications::Request::message() const
+{
+    return m_message;
+}
+
+Notifications::Category Notifications::Request::category() const
+{
+    return m_category;
+}
+
+Notifications::Urgency Notifications::Request::urgency() const
+{
+    return m_urgency;
+}
+
+Notifications::Severity Notifications::Request::severity() const
+{
+    return m_severity;
+}
+
+BitTorrent::InfoHash Notifications::Request::torrent() const
+{
+    return m_torrentHash;
+}
+
+const Notifications::Request::ActionsMap &Notifications::Request::actions() const
+{
+    return m_actions;
+}
+
+int Notifications::Request::timeout() const
+{
+    return m_timeout;
+}
+
+const QVariant & Notifications::Request::userData() const
+{
+    return m_userData;
+}
+
+Notifications::Request::this_type &Notifications::Request::setTitle(const QString &title)
+{
+    m_title = title;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::setMessage(const QString &message)
+{
+    m_message = message;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::setCategory(Category category)
+{
+    m_category = category;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::setUrgency(Urgency urgency)
+{
+    m_urgency = urgency;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::setSeverity(Severity severity)
+{
+    m_severity = severity;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::setTorrent(BitTorrent::InfoHash &infoHash)
+{
+    m_torrentHash = infoHash;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::setTorrent(BitTorrent::TorrentHandle *torrent)
+{
+    m_torrentHash = torrent ? torrent->info().hash() : BitTorrent::InfoHash();
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::setTimeout(int timeout)
+{
+    m_timeout = timeout;
+    return *this;
+}
+
+Notifications::Request::this_type & Notifications::Request::setUserData(const QVariant& data)
+{
+    m_userData = data;
+    return *this;
+}
+
+Notifications::Request::this_type &Notifications::Request::addAction(const QString &id, const QString &label)
+{
+    if (!m_actions.contains(id))
+        m_actions[id] = label;
+    else
+        qDebug() << "Action with id " << id << " is alrady exists. Skipping";
+    return *this;
+}
+
+void Notifications::Request::exec() const &
+{
+    Manager::instance()->notifier()->showNotification(*this);
+}
+
+void Notifications::Request::exec() &&
```

Comment:
The notifier object may need to store notification request, and hence move semantic may be of use.